File: src/main/java/chapter4/transporm/ConcatMapExample.java
package chapter4.transporm;

import common.CommonUtils;
import common.Log;
import io.reactivex.Observable;

import java.util.concurrent.TimeUnit;

public class ConcatMapExample {

    public void marbleDiagram() {
        CommonUtils.exampleStart();
        String[] balls = {"1", "3", "5"};
        Observable<String> source = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map(Long::intValue)
                .map(idx -> balls[idx])
                .take(balls.length)
                .concatMap(ball -> Observable.interval(200L, TimeUnit.MILLISECONDS)
                        .map(notUsed -> ball + "<>")
                        .take(2)
                );
        source.subscribe(Log::it);
        CommonUtils.sleep(2000);
    }

    public void interleaving() {
        CommonUtils.exampleStart();
        String[] balls = {"1","3","5"};
        Observable<String> source = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map(Long::intValue)
                .map(idx -> balls[idx])
                .take(balls.length)
                .flatMap(ball -> Observable.interval(200L, TimeUnit.MILLISECONDS)
                        .map(notUsed -> ball + "<>")
                        .take(2)
                );
        source.subscribe(Log::it);
        CommonUtils.sleep(2000);
    }

    public static void main(String[] args) {
        ConcatMapExample demo = new ConcatMapExample();
        demo.marbleDiagram();
        demo.interleaving();
    }

}

File: src/main/kotlin/chapter4/transporm/ConcatMapExample_.kt
package chapter4.transporm

import common.CommonUtils_
import common.Log_
import io.reactivex.Observable
import java.util.concurrent.TimeUnit

class ConcatMapExample_ {

    fun marbleDiagram() {
        CommonUtils_.exampleStart()
        val balls = arrayOf("1", "3", "5")
        val source = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map { it.toInt() }
                .map { balls[it] }
                .take(balls.size.toLong())
                .concatMap { ball ->
                    Observable.interval(200L, TimeUnit.MILLISECONDS)
                            .map { "$ball<>" }
                            .take(2)
                }
        source.subscribe { Log_.it(it) }
        CommonUtils_.sleep(2000)
    }

    fun interleaving() {
        CommonUtils_.exampleStart()
        val balls = arrayOf("1", "3", "5")
        val source = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map { it.toInt() }
                .map { balls[it] }
                .take(balls.size.toLong())
                .flatMap { ball ->
                    Observable.interval(200L, TimeUnit.MILLISECONDS)
                            .map { "$ball<>" }
                            .take(2)
                }
        source.subscribe { Log_.it(it) }
        CommonUtils_.sleep(2000)
    }

}


fun main(args: Array<String>) {
    val demo = ConcatMapExample_()
    demo.marbleDiagram()
    demo.interleaving()
}
